### PR TITLE
Add global phase gate.

### PR DESCRIPTION
**Pull request by sergeisakov**

### CODE REVIEW

**Comment by 95-martin-orion** [comment]
Before I dig into review: this PR impacts more of qsim than I would have expected. Is this primarily due to adding support for zero-qubit gates, or were there other complications that expanded the change?

**Comment by sergeisakov** [comment]
Yes, this is due to adding support for zero-qubit gates. Perhaps the only non-trivial change in this PR is addition to the fusers.

File: lib/fuser_basic.h [hunk 1]
```diff
@@ -212,6 +220,8 @@ class BasicGateFuser final : public Fuser<IO, Gate> {
           gates_lat[gate.qubits[0]].push_back(&gate);
           gates_lat[gate.qubits[1]].push_back(&gate);
           gates_seq.push_back(&gate);
+        } else {
```
Comment (95-martin-orion) [inline lib/fuser_basic.h:223]: Is there any case where this `else` triggers for gates with >2 qubits?

File: lib/simulator_custatevec.h [hunk 2]
```diff
@@ -80,23 +98,30 @@ class SimulatorCuStateVec final {
...
                            const fp_type* matrix, State& state) const {
-    std::vector<int32_t> control_bits;
-    control_bits.reserve(cqs.size());
-
-    for (std::size_t i = 0; i < cqs.size(); ++i) {
-      control_bits.push_back((cmask >> i) & 1);
-    }
+    if (qs.size() == 0) {
+      IO::errorf(
+          "error: controlled global phase gate is not implemented %s %d\n",
```
Comment (95-martin-orion) [inline lib/simulator_custatevec.h:103]: Is this intended for this PR, planned for a future PR, or not on the roadmap?

**Review by 95-martin-orion — APPROVED** [review]

File: lib/fuser_basic.h (hunk 1 above)
Comment (sergeisakov) [reply lib/fuser_basic.h:223]: I think there is no such a case. The `else` in line 206 triggers for gates with >2 qubits

File: lib/simulator_custatevec.h (hunk 2 above)
Comment (sergeisakov) [reply lib/simulator_custatevec.h:103]: It is not on the roadmap.

**Comment by rht** [comment]
Thank you so much for the PR!

**Pull request merged**